### ISSUE TITLE
enabling 'verify_peer'  => true, connection fails: "IMAP Error: Login failed ... Could not connect ... Unknown reason" ?

### ISSUE THREAD

**Issue by pgnd** [roundcube/roundcubemail#7514]
TL;DR

1.   ssl peer verification to dovecot works with `openssl s_client`
2.   roundcube -> dovecot, OK -- if "verify_peer => false;"
3.   roundcube -> dovecot, FAILS -- if "verify_peer => true;", with "Unknown reason" error

i'm running
```
	roundcube 1.4.7

	git log -n1
	      1 commit cdbefb54e2bebbc61e5fb081c7d1038d884743cf (HEAD, tag: 1.4.7)
	      2 Author: Thomas Bruederli <thomas@roundcube.net>
	      3 Date:   Sat Jul 4 12:32:28 2020 +0200
	      4
	      5     Bump version to 1.4.7

	dovecot --version
		2.3.10.1 (a3d0e1171)


	php -v
		PHP 7.4.8 (cli) (built: Jul  9 2020 08:57:23) ( NTS )
		Copyright (c) The PHP Group
		Zend Engine v3.4.0, Copyright (c) Zend Technologies
		    with Zend OPcache v7.4.8, Copyright (c), by Zend Technologies

	nginx -v
		nginx version: nginx/1.19.0 (local build)
```

on
```
	grep PRETTY /etc/os-release
		PRETTY_NAME="Fedora 32 (Thirty Two)"
```

dovecot's set up & functioning; IMAP works fine with Thunderbird.

connection to dovecot, with peer verification of dovecot's SSL cert, works & identifies "Subject:" & "DNS:" as,
```
	openssl s_client \
	 -4 \
	 -bind 10.12.1.13 \
	 -connect internal.example.com:993 \
	 -verify +9 \
	 -verify_return_error \
	 -verify_hostname internal.example.com \
	 -verify_depth 9 \
	 -ciphersuites TLS_CHACHA20_POLY1305_SHA256 \
	 -cipher ECDHE-ECDSA-CHACHA20-POLY1305 \
	 -cert   /sec/ssl/roundcube.client.ec.crt \
	 -key    /sec/ssl/roundcube.ec.client.key \
	 -CAfile /sec/ssl/ca.chain.crt \
	 -showcerts 2>&1 \
	| openssl x509 -noout -text \
	| egrep 'Subject: |DNS:'


		Subject: C = US, ST = CA, L = HQ, O = example.com, OU = my_CA, CN = internal.example.com, emailAddress = ssl@example.com
		DNS:internal.example.com, DNS:www.internal.example.com, DNS:localhost
```

roundcube, config'd similarly, but WITHOUT peer verification,
```
	$config['debug_level'] = 1;
	$config['imap_debug']     = true;
	$config['default_host'] = 'ssl://internal.example.com:993';
	$config['default_port'] = 993;

	$config['imap_conn_options'] = array(
	 'ssl' => array(
	  'verify_peer'       => true,
	  'verify_peer_name'  => false,
	  'peer_name'         => 'internal.example.com',
	  'verify_depth'      => 9,
	  'allow_self_signed' => true,
	  'SNI_enabled'       => true,
	  'ciphers'    => 'TLS_CHACHA20_POLY1305_SHA256:ECDHE-ECDSA-CHACHA20-POLY1305',
	  'local_cert' => '/sec/ssl/roundcube.client.ec.crt',
	  'local_pk'   => '/sec/ssl/roundcube.ec.client.key',
	  'cafile'     => '/sec/ssl/ca.chain.crt'
	 ),
	);
```

also connects to the dovecot store without error, & works perfectly with full access/functionality @ my IMAP store.

but, if i toggle rouncube's peer verification ==> ON,
```
-	  'verify_peer'       => false,
+	  'verify_peer'       => true,
```

it FAILs to connect,

> IMAP Error: Login failed ... Could not connect ... Unknown reason in .../roundcubemail/program/lib/Roundcube/rcube_imap.php on line 200 (POST /?_task=login&_action=login)

_debug_ logs show only,
```
	tail -f /var/log/dovecot/* /var/log/roundcube/*

		[28-Jul-2020 02:45:32 +0000]: <uobpprkr> [C71D] Connecting to ssl://internal.example.com:993...

		==> /var/log/roundcube/errors.log <==
		[28-Jul-2020 02:45:32 +0000]: <uobpprkr> IMAP Error: Login failed for testuser@example-2.com against internal.example.com from 10.12.1.7. Could not connect to ssl://internal.example.com:993: Unknown reason in /usr/local/src/roundcubemail/program/lib/Roundcube/rcube_imap.php on line 200 (POST /?_task=login&_action=login)

		==> /var/log/dovecot/dovecot-info.log <==
		2020-07-27 19:45:32 imap-login: Info: Disconnected (disconnected before auth was ready, waited 0 secs): user=<>, rip=10.12.1.13, lip=10.12.1.13, TLS handshaking: Connection closed

		==> /var/log/dovecot/dovecot-debug.log <==
		2020-07-27 19:45:32 auth: Debug: Loading modules from directory: /usr/lib64/dovecot/auth
		2020-07-27 19:45:32 auth: Debug: Module loaded: /usr/lib64/dovecot/auth/lib20_auth_var_expand_crypt.so
		2020-07-27 19:45:32 auth: Debug: Module loaded: /usr/lib64/dovecot/auth/libdriver_mysql.so
		2020-07-27 19:45:32 auth: Debug: Module loaded: /usr/lib64/dovecot/auth/libdriver_sqlite.so
		2020-07-27 19:45:32 auth: Debug: Read auth token secret from /run/dovecot//auth-token-secret.dat
		2020-07-27 19:45:32 auth: Debug: passwd-file /usr/local/etc/dovecot/sec/users.conf: Read 10 users in 0 secs
```

**Comment by alecpl** [comment]
I'm afraid we cannot help you. This will be PHP setup issue, not Roundcube issue.

**Comment by pgnd** [comment]
done -> https://bugs.php.net/bug.php?id=79909

we'll see what we learn there ...

thx.

**Comment by pgnd** [comment]
@alecpl 

i've re-installed RC,
```
	git log -n1
	      1 commit cdbefb54e2bebbc61e5fb081c7d1038d884743cf (HEAD, tag: 1.4.7)
	      2 Author: Thomas Bruederli <thomas@roundcube.net>
	      3 Date:   Sat Jul 4 12:32:28 2020 +0200
	      4
	      5     Bump version to 1.4.7
```

connecting to a dovecot backend,
```
	dovecot --version
		2.3.10.1 (a3d0e1171)
```

with
```
	php -v
		PHP 7.4.8 (cli) (built: Jul  9 2020 08:57:23) ( NTS )
		Copyright (c) The PHP Group
		Zend Engine v3.4.0, Copyright (c) Zend Technologies
		    with Zend OPcache v7.4.8, Copyright (c), by Zend Technologies
```

connecting directly with a php stream_socket_client test script to the backend, with SSL vert verification == ON,
```
	test.php
		<?php

		$config = array();

		$prot     = 'ssl';
		$host     = 'example.com';
		$port     = '993';

		$username = 'me@example2.com';
		$password = 'changeme';

		$ssl_params = array(
			'ssl' => array(
				'verify_peer'       => true,
				'verify_depth'      => 2,
				'verify_peer_name'  => true,
				'peer_name'         => 'internal.example.com',
				'allow_self_signed' => true,
				'SNI_enabled'       => true,
				'ciphers'           => 'TLS_CHACHA20_POLY1305_SHA256:ECDHE-ECDSA-CHACHA20-POLY1305',
				'local_cert' => '/srv/ssl/mail.client.EC.crt',
				'local_pk'   => '/srv/ssl/mail.client.EC.key',
				'cafile'     => '/srv/ssl/myCA.CHAIN.crt'
			),
		);
		$ssl_context = stream_context_create($ssl_params);
		$socket = stream_socket_client(
			$prot . '://' . $host . ':' . $port,
			$errno,
			$errstr,
			30,
			STREAM_CLIENT_CONNECT,
			$ssl_context
		);
		echo fread($socket, 16384) . "\n";
		fclose($socket);
		?>
```

works, as expected
```
	php test.php
		* OK [CAPABILITY IMAP4rev1 SASL-IR LOGIN-REFERRALS ID ENABLE IDLE SPECIAL-USE LITERAL+ AUTH=PLAIN] Dovecot ready.
```

switching to RC, with config.inc.php, 1st withOUT cert verify,
```
	...
	$config['default_host'] = 'ssl://example.com';
	$config['default_port'] = 993;
	$config['imap_conn_options'] = array(
		'ssl' => array(
			'verify_peer'       => false,
			'verify_depth'      => 2,
			'verify_peer_name'  => true,
			'peer_name'         => 'internal.example.com',
			'allow_self_signed' => true,
			'SNI_enabled'       => true,
			'ciphers'           => 'TLS_CHACHA20_POLY1305_SHA256:ECDHE-ECDSA-CHACHA20-POLY1305',
			'local_cert' => '/srv/ssl/mail.client.EC.crt',
			'local_pk'   => '/srv/ssl/mail.client.EC.key',
			'cafile'     => '/srv/ssl/myCA.CHAIN.crt'
		),
	);
	...
```

login @ RC UI works perfectly,
```
	...
	==> /var/log/nginx/roundcubemail/imap.log <==
	[14-Aug-2020 21:03:31]: <2c99kdhj> [282B] S: A0002 OK [CAPABILITY IMAP4rev1 SASL-IR LOGIN-REFERRALS ID ENABLE IDLE SORT SORT=DISPLAY THREAD=REFERENCES THREAD=REFS THREAD=ORDEREDSUBJECT MULTIAPPEND URL-PARTIAL CATENATE UNSELECT CHILDREN NAMESPACE UIDPLUS LIST-EXTENDED I18NLEVEL=1 CONDSTORE QRESYNC ESEARCH ESORT SEARCHRES WITHIN CONTEXT=SEARCH LIST-STATUS BINARY MOVE SNIPPET=FUZZY PREVIEW=FUZZY STATUS=SIZE SPECIAL-USE LITERAL+ NOTIFY SPECIAL-USE QUOTA ACL RIGHTS=texk] Logged in
	...
```

RC's fully functional.

next, turning verify == ON in RC config,
```
-			'verify_peer'       => false,
+			'verify_peer'       => true,
```

so that the SSL connection params for the direct-php & RC cases match exactly,

with verify == ON, login @ RC UI FAILs to connect,
```
	==> /var/log/nginx/roundcubemail/imap.log <==
	[14-Aug-2020 21:05:26]: <ehnquuum> [5643] Connecting to ssl://example.com:993...

	==> /var/log/nginx/roundcubemail/errors.log <==
	[14-Aug-2020 21:05:26]: <ehnquuum> IMAP Error: Login failed for me@example2.com against example.com from 10.1.1.7. Could not connect to ssl://example.com:993: Unknown reason in /srv/roundcubemail/program/lib/Roundcube/rcube_imap.php on line 200 (POST /?_task=login&_action=login)

	==> /var/log/dovecot/dovecot-debug.log <==
	2020-08-14 14:05:26 auth: Debug: auth client connected (pid=13557)

	==> /var/log/dovecot/dovecot-info.log <==
	2020-08-14 14:05:26 imap-login: Info: Disconnected (no auth attempts in 0 secs): user=<>, rip=127.0.0.1, lip=127.0.0.1, TLS handshaking: Connection closed
```

> This will be PHP setup issue, not Roundcube issue.

can you clarify what specific changed or addition PHP setup is needed in Roundcube config? As, the observations, from above,

 (1)	php connect to imap, WITH    ssl cert verification WORKS
    (fyi, as does openssl s_client ... with verification )
 (2)	RC  connect to imap, WITHOUT ssl cert verification WORKS
 (3)	RC  connect to imap, WITH    ssl cert verification FAILS

are completely reproducible.

**Comment by davidebeatrici** [comment]
Please note that this issue can be encountered with the default configuration. The workaround consists in adding the following to `config.inc.php`:

```php
$config['imap_conn_options']['ssl']['verify_peer_name'] = false;
```

**Comment by n0099** [comment]
or
```php
$config['imap_conn_options']['ssl']['peer_name'] = YOUR_SubjectAltName_OF_CERT;
```
